modules/tftp/http_downloads.rb

```diff
@@ -0,0 +1,297 @@
+require 'uri'
+require 'securerandom'
+require 'digest'
+require 'json'
+require 'concurrent'
+
+module Proxy::TFTP
+  class HttpError < RuntimeError
+    attr_reader :code, :message
+
+    def initialize(code, message)
+      @code = code
+      @message = message
+    end
+  end
+
+  class HttpDownload
+    class Status
+      attr_accessor :last_error, :timestamp, :stopped
+
+      def initialize(now = Time.now)
+        @m = Mutex.new
+        @file_length = 0
+        @downloaded = 0
+        @timestamp = now
+        @stopped = false
+      end
+
+      def file_length
+        @m.synchronize { @file_length }
+      end
+
+      def downloaded
+        @m.synchronize { @downloaded }
+      end
+
+      def file_length=(length)
+        @m.synchronize { @file_length = length }
+      end
+
+      def update_downloaded(chunk_length)
+        @m.synchronize { @downloaded += chunk_length }
+      end
+
+      def progress
+        (fl = file_length) == 0 ? 0 : ((downloaded.to_f/fl) * 100).round(2).to_i
+      end
+
+      def reset_error
+        @last_error = nil
+      end
+
+      def stopped?
+        @stopped
+      end
+
+      def to_hash
+        {:file_length => file_length, :downloaded => downloaded, :progress => progress, :timestamp => timestamp, :last_error => last_error}
+      end
+    end
+
+    class Downloader
+      ATTEMPT_DELAY = 10
+
+      def initialize(parsed_url, dst_path, status)
+        @stop = false
+        @status = status
+        @parsed_url = parsed_url
+        @dst_path = dst_path
+      end
+
+      def start(start_position = 0)
+        Net::HTTP.start(@parsed_url.host, @parsed_url.port, :use_ssl => @parsed_url.scheme == 'https') do |http|
+          get = Net::HTTP::Get.new(@parsed_url, (range_header(start_position) if start_position != 0))
+          http.request(get) do |response|
+            raise ::Proxy::TFTP::HttpError.new(response.code, response.message) unless response.is_a?(Net::HTTPSuccess)
+            @status.file_length = response['content-length'].to_i rescue 0
+            @status.update_downloaded(start_position) if response.code == '206'
+            open(@dst_path, (response.code == '206' ? 'ab' : 'wb')) do |io|
+              if @stop
+                @status.stopped = true
+                break
+              end
+              response.read_body {|b| @status.update_downloaded(b.length); io.write(b)}
+            end
+          end
```

Comment:
> I don't see any IO seek in the target file in this method, how this code can correctly resume downloads? This cannot work.

'Append' mode is used on successful resumes, which appends to the end of an existing file, no need for a seek call.

The code dealing with actually downloading stuff is pretty small (and well tested). But it gives me ability to respond to different modes of failures (transient failures vs. permanent), which is very useful. 

> Isn't synchronized queue better approach than mutex with synchronized blocks everywhere?

I would still need synchronization, as there's a couple of pieces of shared state (id to download and url to download mappings; synchronization around destination dir creation, etc).